### DISCUSSION TITLE
Compress Simplified Chinese slow

### DISCUSSION THREAD

**Discussion by WangJietao123** [fonttools/fonttools#2876] [Q&A]
Hello. I want to know that the compression speed of subset compression in Simplified Chinese is very slow. Is there any good optimization scheme or has anyone solved this problem in the past?

**Comment by WangJietao123** [comment]
I wrote an api interface to call this service, but more than 60s will exceed the timeout time of nginx

**Comment by justvanrossum** [comment]
If it's about subsetting, you should look into the HarfBuzz subsetter, which is much faster:
- https://docs.google.com/presentation/d/1RMkpikAUmdIXaz3eEDYFETWfAwnT0Vw8_oC8Ait5xqs/edit#slide=id.g41919619d9_0_65
(But I can't find a link for how to install and how to operate :( )

If it's about woff2-compressing an already subsetted font: woff2 compression can be quite slow for big fonts, but you could try this tool instead: https://github.com/google/woff2

**Reply by WangJietao123** [reply]
I extract simplified Chinese from a complete font（subset）. I don't know why the compression time is so different when the file size is not so different? (Want to know the reason) For example, Source Han Sans takes 80 seconds (some also take dozens of seconds, about one sixth of the total). But most other files only need 1-2s.

**Reply by justvanrossum** [reply]
Can you show the exact commands you are using to compare the timing of these cases? It is not clear to me what you are doing.

**Reply by WangJietao123** [reply]
```python
import os, uuid
from pathlib import Path
from fontTools.subset import main as subsetFunc

def compressFont(
    fontLocalPath: str,
    fontType: str,
    unicodes: str,
    subset: str
):
    outputFileName = str(uuid.uuid4()) + '.' + fontType
    outputFilePath = '/tmp/font_output/' + outputFileName

    parameter = [
        fontLocalPath,
        "--unicodes=%s" % unicodes,
        "--layout-features=*",
        "--name-IDs=*",
        "--name-languages=*",
        "--flavor=%s" % fontType,
        "--output-file=%s" % outputFilePath,
    ]
    if subset:
        for pm in subset:
            parameter.append(pm)
    
    subsetFunc(parameter)

    return outputFilePath

__all__ = [
    'compressFont',
    'outputFileDir',
]
```

**Reply by justvanrossum** [reply]
With what kind of inputs do you call that function? What inputs do you give to get the 80 seconds timing vs. 1-2 seconds? Are inputs OTF or TTF, is output woff2? How big are the input files? How many characters does the subset contain? Etc.

Btw. my original comment still holds: if you need this to be faster, you are very likely to get better results using the HarfBuzz subsetter.

**Reply by WangJietao123** [reply]
input ttf output woff     The file size in 80 seconds is 36mb and have It's about 65535 characters, and the file size in 2 seconds is 13mb,parameters are the same.
I will try HarfBuzz subsetter.THX

**Reply by justvanrossum** [reply]
What are the sizes of the resulting files?

**Reply by WangJietao123** [reply]
13.1mb

> What are the sizes of the resulting files?
